src/psyclone/psyir/nodes/psy_data_node.py

```diff
@@ -216,7 +162,111 @@ def __init__(self, ast=None, children=None, parent=None, options=None):
+            (``prefix_PSyDataType``) - a "_" will be added automatically. \
+            It defaults to "", which means the module name used will just be \
+            ``psy_data_mod``, and the data type ``PSyDataType``.
+        :param (str,str) options["region_name"]: an optional name to use for \
+            this PSyDataNode, provided as a 2-tuple containing a module name \
+            followed by a local name. The pair of strings should uniquely \
+             identify a region unless aggregate information is required \
+            (and is supported by the runtime library).
+
+        :raises TypeError: if the supplied children or symbol table are not \
+            of the correct type.
+
+        '''
+        if not isinstance(children, list):
+            raise TypeError("Error in PSyDataNode.create(). The 'children' "
+                            "argument must be a list (of PSyIR nodes) but got "
+                            "'{0}'".format(type(children).__name__))
+        if children and not all(isinstance(child, Node) for child in children):
+            raise TypeError(
+                "Error in PSyDataNode.create(). The 'children' argument must "
+                "be a list of PSyIR nodes but it contains: {0}".format(
+                    [type(child).__name__ for child in children]))
+        if not isinstance(symbol_table, SymbolTable):
+            raise TypeError(
+                "Error in PSyDataNode.create(). The 'symbol_table' argument "
+                "must be an instance of psyir.symbols.SymbolTable but got "
+                "'{0}'.".format(type(symbol_table).__name__))
+
+        data_node = cls(ast=ast, options=options)
+
+        # Ensure that we have a container symbol for the API access
+        fortran_module = data_node.add_psydata_class_prefix(
+            data_node.fortran_module)
+        try:
+            csym = symbol_table.lookup_with_tag(fortran_module)
+        except KeyError:
+            # The tag doesn't exist which means that we haven't already added
+            # this Container as part of a PSyData transformation.
+            csym = ContainerSymbol(fortran_module)
+            symbol_table.add(csym, tag=fortran_module)
+
+        # A PSyData node always contains a Schedule
+        # TODO 435 we can probably get rid of _insert_schedule() once we
+        # do away with the ast argument?
+        # pylint: disable=protected-access
+        sched = data_node._insert_schedule(children=children, ast=ast)
+        data_node.addchild(sched)
+
+        # The use statement that will be inserted by the update() method. Any
+        # use of a module of the same name that doesn't match this will result
+        # in a NotImplementedError at code-generation time.
+        # TODO #435 remove this when removing the update() method
+        names = [data_node.add_psydata_class_prefix(symbol.name) for symbol in
+                 PSyDataNode.symbols]
```

Comment:
s/PSyDataNode/self/ ? Or we explicitly don't want to allow subclasses to redefine this?

Reply:
I wasn't imagining that they would but of course they could. Changed.